File: aeroshield/core/protocol.py
"""UART binary protocol between host PC and ESP32 (KTR 4.3.2)."""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
from typing import Optional


SOF = 0xAA
EOF = 0xFF
# [AA][MOD][PAN_H][PAN_L][TILT_H][TILT_L][FIRE]
# [IDLE_H][IDLE_L][FIRE_H][FIRE_L][SPIN_CS][CHK][FF]
CMD_LEN = 14
STATUS_LEN = 6
ESC_MIN_US = 1000
ESC_MAX_US = 2000  # full ESC range; panel can set freely for now


class ModCode(IntEnum):
    MANUAL = 0
    SEMI = 1
    AUTO = 2
    ESTOP = 3
    HOME = 4


class StatusCode(IntEnum):
    OK = 0
    LIMIT = 1
    FAILSAFE = 2
    ESTOP = 3
    HOME = 4
    BUSY = 5


def xor_checksum(data: bytes) -> int:
    c = 0
    for b in data:
        c ^= b
    return c & 0xFF


def encode_i16(value: int) -> tuple[int, int]:
    value = max(-32768, min(32767, int(value)))
    if value < 0:
        value += 65536
    return (value >> 8) & 0xFF, value & 0xFF


def decode_i16(hi: int, lo: int) -> int:
    value = ((hi & 0xFF) << 8) | (lo & 0xFF)
    if value >= 32768:
        value -= 65536
    return value


def clamp_esc_us(us: int) -> int:
    return max(ESC_MIN_US, min(ESC_MAX_US, int(us)))


@dataclass
class CommandPacket:
    mod: int = ModCode.MANUAL
    pan: int = 0
    tilt: int = 0
    fire: int = 0
    esc_idle_us: int = 1060
    esc_fire_us: int = 1200
    fire_spin_ms: int = 450

    def to_bytes(self) -> bytes:
        pan_h, pan_l = encode_i16(self.pan)
        tilt_h, tilt_l = encode_i16(self.tilt)
        idle = clamp_esc_us(self.esc_idle_us)
        fire_us = clamp_esc_us(self.esc_fire_us)
        if fire_us < idle:
            fire_us = idle
        idle_h, idle_l = encode_i16(idle)
        fire_h, fire_l = encode_i16(fire_us)
        spin_cs = max(1, min(255, int(round(self.fire_spin_ms / 10.0))))
        body = bytes(
            [
                SOF,
                int(self.mod) & 0xFF,
                pan_h,
                pan_l,
                tilt_h,
                tilt_l,
                1 if self.fire else 0,
                idle_h,
                idle_l,
                fire_h,
                fire_l,
                spin_cs & 0xFF,
            ]
        )
        checksum = xor_checksum(body[1:])  # MOD..SPIN
        return body + bytes([checksum, EOF])


@dataclass
class StatusPacket:
    status: int = StatusCode.OK
    limit_pan: int = 0
    limit_tilt: int = 0

    @classmethod
    def from_bytes(cls, raw: bytes) -> Optional["StatusPacket"]:
        if len(raw) != STATUS_LEN:
            return None
        if raw[0] != SOF or raw[-1] != EOF:
            return None
        payload = raw[1:4]
        checksum = raw[4]
        if xor_checksum(payload) != checksum:
            return None
        return cls(status=raw[1], limit_pan=raw[2], limit_tilt=raw[3])

    def to_bytes(self) -> bytes:
        payload = bytes(
            [int(self.status) & 0xFF, int(self.limit_pan) & 0xFF, int(self.limit_tilt) & 0xFF]
        )
        return bytes([SOF]) + payload + bytes([xor_checksum(payload), EOF])


class PacketParser:
    """Incremental parser for ESP32 status frames."""

    def __init__(self) -> None:
        self._buf = bytearray()

    def feed(self, data: bytes) -> list[StatusPacket]:
        self._buf.extend(data)
        packets: list[StatusPacket] = []
        while True:
            try:
                start = self._buf.index(SOF)
            except ValueError:
                self._buf.clear()
                break
            if start > 0:
                del self._buf[:start]
            if len(self._buf) < STATUS_LEN:
                break
            chunk = bytes(self._buf[:STATUS_LEN])
            pkt = StatusPacket.from_bytes(chunk)
            if pkt is None:
                del self._buf[0]
                continue
            del self._buf[:STATUS_LEN]
            packets.append(pkt)
        return packets

File: firmware/aeroshield_esp32/protocol.cpp
#include "protocol.h"

namespace proto {

uint8_t xorChecksum(const uint8_t* data, size_t len) {
  uint8_t c = 0;
  for (size_t i = 0; i < len; ++i) {
    c ^= data[i];
  }
  return c;
}

int16_t decodeI16(uint8_t hi, uint8_t lo) {
  uint16_t raw = (static_cast<uint16_t>(hi) << 8) | lo;
  if (raw >= 32768) {
    return static_cast<int16_t>(static_cast<int32_t>(raw) - 65536);
  }
  return static_cast<int16_t>(raw);
}

namespace {

enum ParseState : uint8_t {
  WAIT_SOF,
  READ_BODY,
};

ParseState state = WAIT_SOF;
uint8_t buf[CMD_LEN];
size_t idx = 0;

uint16_t clampEscUs(uint16_t us) {
  if (us < ESC_MIN_US) {
    return ESC_MIN_US;
  }
  if (us > ESC_CAP_US) {
    return ESC_CAP_US;
  }
  return us;
}

}  // namespace

bool parseCommandByte(uint8_t byte, CommandPacket& out, bool& complete) {
  complete = false;

  if (state == WAIT_SOF) {
    if (byte == SOF) {
      buf[0] = byte;
      idx = 1;
      state = READ_BODY;
    }
    return false;
  }

  buf[idx++] = byte;
  if (idx < CMD_LEN) {
    return false;
  }

  state = WAIT_SOF;
  idx = 0;

  if (buf[CMD_LEN - 1] != EOF_MARK) {
    return false;
  }

  // MOD..SPIN_CS = 11 bytes starting at buf[1]
  const uint8_t checksum = xorChecksum(buf + 1, 11);
  if (checksum != buf[12]) {
    return false;
  }

  out.mod = buf[1];
  out.pan = decodeI16(buf[2], buf[3]);
  out.tilt = decodeI16(buf[4], buf[5]);
  out.fire = (buf[6] != 0);
  out.escIdleUs = clampEscUs(static_cast<uint16_t>(decodeI16(buf[7], buf[8])));
  out.escFireUs = clampEscUs(static_cast<uint16_t>(decodeI16(buf[9], buf[10])));
  if (out.escFireUs < out.escIdleUs) {
    out.escFireUs = out.escIdleUs;
  }
  out.fireSpinMs = static_cast<uint16_t>(buf[11]) * 10u;
  if (out.fireSpinMs < 50u) {
    out.fireSpinMs = 50u;
  }
  complete = true;
  return true;
}

void buildStatusFrame(const StatusPacket& pkt, uint8_t* out6) {
  const uint8_t payload[3] = {pkt.status, pkt.limitPan, pkt.limitTilt};
  out6[0] = SOF;
  out6[1] = payload[0];
  out6[2] = payload[1];
  out6[3] = payload[2];
  out6[4] = xorChecksum(payload, 3);
  out6[5] = EOF_MARK;
}

}  // namespace proto
